File: Python3/801 - 1000/814 Binary Tree Pruning.py
# 40ms, 47.02%; 13.4MB, 7.25%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pruneTree(self, root: TreeNode) -> TreeNode:
        if not root: return None
        self.prune(root)
        return None if root.val == 0 and root.left is root.right else root

    def prune(self, root: TreeNode):
        if root.left:
            if self.check(root.left):
                root.left = None
            else:
                self.prune(root.left)
        if root.right:
            if self.check(root.right):
                root.right = None
            else:
                self.prune(root.right)

    def check(self, root: TreeNode) -> bool:
        flag = (root.val == 0)
        if root.left:
            flag &= self.check(root.left)
        if root.right:
            flag &= self.check(root.right)
        return flag

File: 801 - 1000/814 Binary Tree Pruning.cc
// 4ms, 100.00%; 9.9MB, 67.71%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
public:
    TreeNode *pruneTree(TreeNode *root)
    {
        if (!root) return nullptr;
        prune(root);
        return (root->val == 0 && root->left == root->right ? nullptr : root);
    }

private:
    void prune(TreeNode *root)
    {
        if (root->left) {
            if (check(root->left)) {
                root->left = nullptr;
            } else {
                prune(root->left);
            }
        }
        if (root->right) {
            if (check(root->right)) {
                root->right = nullptr;
            } else {
                prune(root->right);
            }
        }
    }

    bool check(TreeNode *root)
    {
        bool flag = (root->val == 0);
        if (root->left) flag &= check(root->left);
        if (root->right) flag &= check(root->right);
        return flag;
    }
};
